### PR TITLE
EI Backend Instances fault handling improvements.

### PR DESCRIPTION
**Pull request by tobiasake**
Today we have had many situations where initiation of EI Backend Instances list fails without logging any issues or informing the EI Frontend user Web-UI.

This is an early version,, it will come more changes to this pull-request.

### Description of the Change
Improvement of logging and informing user about what is failing during EI Instance List initiations.

### Sign-off

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.

Signed-off-by:

### CODE REVIEW

File: src/main/java/com/ericsson/ei/frontend/BackEndInformationController.java [hunk 1]
```diff
@@ -42,15 +42,15 @@
...
     @RequestMapping(value = "/backend", method = RequestMethod.GET)
     public ResponseEntity<String> getInstances(Model model, HttpServletRequest request) {
-        LOG.debug("Recieved request for instances.");
+        LOG.debug("Recieved request for retrieving EI Backend instances list.");
         ResponseEntity<String> response = backEndInfoContUtils.handleRequestForInstances(request);
         return response;
     }
 
     @CrossOrigin
     @RequestMapping(value = "/backend", method = RequestMethod.PUT)
```
Comment (a-breid-ericsson) [inline src/main/java/com/ericsson/ei/frontend/BackEndInformationController.java]: This method is deprecated.

File: src/main/java/com/ericsson/ei/frontend/config/CheckEIConfigurations.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+                + "ei.frontend.service.host: " + env.getProperty("ei.frontend.service.host") + "\n"
+                + "ei.frontend.context.path: " + env.getProperty("ei.frontend.context.path") + "\n"
+                + "ei.use.secure.http.frontend: " + env.getProperty("ei.use.secure.http.frontend") + "\n"
+                + "ei.backend.instances.filepath: " + env.getProperty("ei.backend.instances.filepath") + "\n"
+                + "ei.backend.instances.list.json.content: " + env.getProperty("ei.backend.instances.list.json.content") + "\n"
+                + "ei.eiffel.documentation.urls: " + env.getProperty("ei.eiffel.documentation.urls") + "\n"
+                + "logging.level.root: " + env.getProperty("logging.level.root") + "\n"
+                + "logging.level.org.springframework.web: " + env.getProperty("logging.level.org.springframework.web") + "\n"
+                + "logging.level.com.ericsson.ei: " + env.getProperty("logging.level.com.ericsson.ei") + "\n"
+                + "\nThese properties is only some of the configuraitons, more configurations can have been provided.\n");
```
Comment (a-breid-ericsson) [inline src/main/java/com/ericsson/ei/frontend/config/CheckEIConfigurations.java]: may have been provided.

File: src/main/resources/static/js/subscription.js [hunk 3]
```diff
@@ -30,6 +30,8 @@ jQuery(document).ready(function () {
                 backendStatus = true;
             },
             error: function (XMLHttpRequest, textStatus, errorThrown) {
+                var jsonObj = JSON.parse(XMLHttpRequest.responseText);
```
Comment (a-breid-ericsson) [inline src/main/resources/static/js/subscription.js]: Please check indention here, tabs is used and not spaces.

File: src/main/java/com/ericsson/ei/frontend/BackEndInformationController.java [hunk 4]
```diff
@@ -42,15 +42,15 @@
...
     public ResponseEntity<String> switchBackEndInstance(Model model, HttpServletRequest request) {
-        LOG.debug("Recieved request to switch back end.");
+        LOG.debug("Recieved request to switch EI back end instance.");
```
Comment (e-pettersson-ericsson) [inline src/main/java/com/ericsson/ei/frontend/BackEndInformationController.java]: Please change "back end" to "backend" everywhere so it's consistent in both logging and user documentation.

File: src/main/java/com/ericsson/ei/frontend/config/CheckEIConfigurations.java (hunk 2 above)
Comment (e-pettersson-ericsson) [inline src/main/java/com/ericsson/ei/frontend/config/CheckEIConfigurations.java]: Same spellcheck as in EI backend PR.

File: src/main/java/com/ericsson/ei/frontend/exceptions/EiBackendInstancesException.java [hunk 5]
```diff
@@ -0,0 +1,35 @@
+/*
+   Copyright 2017 Ericsson AB.
```
Comment (e-pettersson-ericsson) [inline src/main/java/com/ericsson/ei/frontend/exceptions/EiBackendInstancesException.java]: Should an new file have 2017 in the copyright?

File: src/main/java/com/ericsson/ei/frontend/utils/BackEndInstanceFileUtils.java [hunk 6]
```diff
@@ -155,14 +157,19 @@ private void setDefaultEiBackendInstance(JsonArray jArray) {
...
      */
     public JsonArray getInstancesFromFile() {
+        LOG.debug("Reading EI Instances File: " + eiInstancesPath);
         try {
             ensureValidFile();
 
             JsonArray inputBackEndInstances = new JsonParser()
                     .parse(new String(Files.readAllBytes(Paths.get(eiInstancesPath)))).getAsJsonArray();
+            if (inputBackEndInstances.isJsonNull() || inputBackEndInstances.size() < 1) {
+                LOG.debug("EI Instance file does not exist or is empty. Returnin an empty EI Instance json list.");
```
Comment (e-pettersson-ericsson) [inline src/main/java/com/ericsson/ei/frontend/utils/BackEndInstanceFileUtils.java]: Missing a g in Returning

File: src/main/java/com/ericsson/ei/frontend/BackEndInformationController.java [hunk 7]
```diff
@@ -59,15 +59,15 @@
     @CrossOrigin
     @RequestMapping(value = "/backend", method = RequestMethod.DELETE)
     public ResponseEntity<String> deleteBackEndInstance(Model model, HttpServletRequest request) {
-        LOG.debug("Recieved request to delete back end.");
+        LOG.debug("Recieved request to delete EI back end instance.");
```
Comment (e-pettersson-ericsson) [inline src/main/java/com/ericsson/ei/frontend/BackEndInformationController.java]: Recieved --> Received . Double check in other log messages also.

File: src/main/java/com/ericsson/ei/frontend/BackEndInformationController.java (hunk 4 above)
Comment (e-pettersson-ericsson) [reply src/main/java/com/ericsson/ei/frontend/BackEndInformationController.java]: You're right, I took a closer look at the documentation and we decided to use back-end and front-end when we refer to the nouns. But my issue here is that the comments in the code was using different words. Please use the hyphened words from now on.

File: src/main/java/com/ericsson/ei/frontend/BackEndInformationController.java (hunk 1 above)
Comment (tobiasake) [reply src/main/java/com/ericsson/ei/frontend/BackEndInformationController.java]: Ok, can it be removed then? Remove it in a own Pull-Request.

File: src/main/java/com/ericsson/ei/frontend/BackEndInformationController.java (hunk 4 above)
Comment (tobiasake) [reply src/main/java/com/ericsson/ei/frontend/BackEndInformationController.java]: I think "backend" term is commonly used in a technical world. I use always "backend" term when I write the word in code and technical comments and emails. Easier to read and spell the word when its not decoupled, thats my opinion. But the documentation on github io pages can maybe use "back-end" or "back end" format.

File: src/main/java/com/ericsson/ei/frontend/config/CheckEIConfigurations.java (hunk 2 above)
Comment (tobiasake) [reply src/main/java/com/ericsson/ei/frontend/config/CheckEIConfigurations.java]: Its OK with "backend" :)

File: src/main/java/com/ericsson/ei/frontend/BackEndInformationController.java (hunk 4 above)
Comment (e-pettersson-ericsson) [reply src/main/java/com/ericsson/ei/frontend/BackEndInformationController.java]: I agree the "backend" is easier to write, and I would prefer using this only.  I'll approve this PR. I made a PR #203 to fix the markdown documentation that I found, and see what @a-breid-ericsson thinks about it :)

Comment (tobiasake) [reply src/main/java/com/ericsson/ei/frontend/BackEndInformationController.java]: Yes, I have added my opinions in that PR as well :)

File: src/main/java/com/ericsson/ei/frontend/config/CheckEIConfigurations.java (hunk 2 above)
Comment (tobiasake) [reply src/main/java/com/ericsson/ei/frontend/config/CheckEIConfigurations.java]: done.

File: src/main/resources/static/js/subscription.js (hunk 3 above)
Comment (tobiasake) [reply src/main/resources/static/js/subscription.js]: no, its correct number of spaces.

File: src/main/java/com/ericsson/ei/frontend/exceptions/EiBackendInstancesException.java (hunk 5 above)
Comment (tobiasake) [reply src/main/java/com/ericsson/ei/frontend/exceptions/EiBackendInstancesException.java]: done

File: src/main/java/com/ericsson/ei/frontend/utils/BackEndInstanceFileUtils.java (hunk 6 above)
Comment (tobiasake) [reply src/main/java/com/ericsson/ei/frontend/utils/BackEndInstanceFileUtils.java]: done

File: src/main/java/com/ericsson/ei/frontend/BackEndInformationController.java (hunk 7 above)
Comment (tobiasake) [reply src/main/java/com/ericsson/ei/frontend/BackEndInformationController.java]: done.

File: src/functionaltest/java/com/ericsson/ei/frontend/pageobjects/SubscriptionPage.java [hunk 8]
```diff
@@ -181,53 +177,48 @@ public boolean expandButtonExist(String XPath) {
     }
 
     public boolean buttonExistByXPath(String XPath) {
-        // The row indicates weather or not the del / view and edit buttons has moved down to
+        // The row indicates weather or not the del / view and edit buttons has moved
+        // down to
```
Comment (Christoffer-Cortes) [inline src/functionaltest/java/com/ericsson/ei/frontend/pageobjects/SubscriptionPage.java]: This is a really weird line break.

File: src/functionaltest/java/com/ericsson/ei/frontend/pageobjects/SubscriptionPage.java [hunk 9]
```diff
@@ -258,23 +251,27 @@ public boolean clickExpandButtonByXPath(String loc) {
...
     public int countElements(String id) {
-        return driver.findElements(By.id(id)).size();
+        return driver.findElements(By.id(id))
+                     .size();
     }
 
     public void clickViewButtonByXPath(String XPath) {
-        // The row indicates weather or not the del / view and edit buttons has moved down to
+        // The row indicates weather or not the del / view and edit buttons has moved
+        // down to
```
Comment (Christoffer-Cortes) [inline src/functionaltest/java/com/ericsson/ei/frontend/pageobjects/SubscriptionPage.java]: Another odd linebreak.

File: src/main/java/com/ericsson/ei/frontend/EIRequestsController.java [hunk 10]
```diff
@@ -73,7 +74,14 @@
             "/queryAggregatedObject", "/queryMissedNotifications", "/query",
             "/rules/rule-check/testRulePageEnabled" }, method = RequestMethod.GET)
     public ResponseEntity<String> getRequests(Model model, HttpServletRequest incomingRequest) {
-        String eiRequestUrl = eiRequestsControllerUtils.getEIRequestURL(incomingRequest);
+        String eiRequestUrl;
```
Comment (Christoffer-Cortes) [inline src/main/java/com/ericsson/ei/frontend/EIRequestsController.java]: From what I can see there are 4 identical code blocks added here. Maybe it's possible to put them in a method.

File: src/functionaltest/java/com/ericsson/ei/frontend/pageobjects/SubscriptionPage.java (hunk 8 above)
Comment (tobiasake) [reply src/functionaltest/java/com/ericsson/ei/frontend/pageobjects/SubscriptionPage.java]: Not my change. Had to rebase to latest code to solve the integrations test issues due to EI Backend changes. and got this into my PR.

File: src/functionaltest/java/com/ericsson/ei/frontend/pageobjects/SubscriptionPage.java (hunk 9 above)
Comment (tobiasake) [reply src/functionaltest/java/com/ericsson/ei/frontend/pageobjects/SubscriptionPage.java]: Not my change. Had to rebase to latest code to solve the integrations test issues due to EI Backend changes. and got this into my PR.

File: src/main/java/com/ericsson/ei/frontend/EIRequestsController.java (hunk 10 above)
Comment (tobiasake) [reply src/main/java/com/ericsson/ei/frontend/EIRequestsController.java]: Thats true,, but its , maybe, only Log message that can be moved out to help method.

File: src/main/java/com/ericsson/ei/frontend/config/CheckEIConfigurations.java [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+                + "ei.frontend.service.host: " + env.getProperty("ei.frontend.service.host") + "\n"
+                + "ei.frontend.context.path: " + env.getProperty("ei.frontend.context.path") + "\n"
+                + "ei.use.secure.http.frontend: " + env.getProperty("ei.use.secure.http.frontend") + "\n"
+                + "ei.backend.instances.filepath: " + env.getProperty("ei.backend.instances.filepath") + "\n"
+                + "ei.backend.instances.list.json.content: " + env.getProperty("ei.backend.instances.list.json.content") + "\n"
+                + "ei.eiffel.documentation.urls: " + env.getProperty("ei.eiffel.documentation.urls") + "\n"
+                + "logging.level.root: " + env.getProperty("logging.level.root") + "\n"
+                + "logging.level.org.springframework.web: " + env.getProperty("logging.level.org.springframework.web") + "\n"
+                + "logging.level.com.ericsson.ei: " + env.getProperty("logging.level.com.ericsson.ei") + "\n"
+                + "\nThese properties is only some of the configuraitons, more configurations may have been provided.\n");
```
Comment (e-pettersson-ericsson) [inline src/main/java/com/ericsson/ei/frontend/config/CheckEIConfigurations.java]: configuraitons -> configurations

File: src/main/java/com/ericsson/ei/frontend/utils/BackEndInstanceFileUtils.java (hunk 6 above)
Comment (e-pettersson-ericsson) [reply src/main/java/com/ericsson/ei/frontend/utils/BackEndInstanceFileUtils.java]: Still missing the G :)

**Pull request merged**